modules/standard/DistributedBag.chpl

```diff
@@ -0,0 +1,1076 @@
+  while operating in that it enables us to manage 'best-case', 'average-case',
+  and 'worst-case' scenarios by making multiple passes over each segment. In the
+  case where there is no oversubscription, the best-case will always be achieved
+  (considering any other conditions are also met), while in the case of oversubscription
+  or, for example a near empty bag, we fall into the 'average-case', etc. Examples
+  of 'best-case' scenarios for a removal would be when a segment is unlocked
+  and contains some elements we can drain, and the 'average-case' would be to find
+  any segment, unlocked or not, that contains elements we can drain, and so on.
+
+  This data structure also employs its own load balancing algorithm, beginning
+  with a best-effort round-robin algorithm such that each task begins searching
+  at another segment (with the added benefit of reducing overall contention), in
+  particular is useful for locally distributing insertion operations. Next it also
+  employs a work-stealing algorithm that horizontally steals a ratio of elements
+  across segments. A segment is considered 'horizontal' if it shares the same
+  segment index within [0, here.maxTaskPar); those other segments likely are also
+  benefitted from local distribution from the round-robin algorithm discussed above.
+  We also steal a ratio of elements, say 25%, because it leaves all other segments
+  with 75% of their work; the more elements they have, the more we take, but the
+  less they have, the less we steal; this also has the added benefit of reducing
+  unnecessary work stealing between segments when the bag is nearly emptied. Furthermore,
+  stealing horizontally ensures that all segments remain filled and that we still
+  achieve parallelism across segments for removal operations. Lastly, we attempt
+  to steal a maximum of `N / sizeof(eltType)`, where N is some size in megabytes
+  (representing how much data can be sent in one network request), which keeps
+  down excessive communication.
+
+  This data structure does not come without flaws; as work stealing is dynamic
+  and triggered on demand, work stealing can still be performed in excess, which
+  dramatically causes a performance drop. Furthermore, a severe imbalance across
+  nodes, such as an unfair distribution of work to a single or small subset of nodes,
+  may also causes an equally severe performance degradation. To circumvent this,
+  we allow the user to determine manually when to 'balance', which will evenly distribute all
+  elements across nodes. This data structure scales in terms of nodes, processors per node,
+  and even work load. The larger the work load, the more data that gets stolen when
+  stealing work, and better locality of elements among segments. As well, to
+  achieve true parallelism, usage of a privatized instance is a requirement, as
+  it avoids the overhead of remotely accessing class fields, bounding performance
+  on communication.

```

Comment:
I'd lean towards mimicking the pattern of record wrapping done by arrays. That would not necessarily mean using OwnedObject. In particular, the main point there is that when the record is destroyed, the Impl instance is deleted, and there's no provision for keeping the instance alive longer if there are record copies of it. In this way it is similar to using a bare class.

It's probably not currently a good idea to try to use OwnedObject for this purpose, as OwnedObject has some issues (to do with lvalue checking and compiler errors at least) that you could easily avoid just by writing your own constructor / destructor.

If you created an array of DistributedBagImpl, you'd still need privatization / replication in that array to avoid communication when getting the local instance ID. I think it's reasonable to use the privatization mechanism, but having a Replicated array field (say) is a pretty reasonable alternative.

About not wanting to use the record as a wide ref - Ben H. recently committed a mechanism that can be used to serialize/deserialize instances. Obviously it might be more than you're able to do in your last week, but you could try to use that to control what happens when your data structure is used on the other side of an `on` statement.

I'm not sure what you mean by forwarding... "If I want to forward only on the privatized instance, is this currently allowed?" I think that's what the array record does, depending on exactly what you mean. Might be best to find me on irc to ask about these.